text/3498-lifetime-capture-rules-2024.md

```diff
@@ -0,0 +1,642 @@
+
+For an example of how TAIT is intended to be used to fix overcapturing, see the section titled "[TAIT as the solution to overcapturing]".
+
+## Summary of problems
+
+In summary, in Rust 2021, the lifetime capture rules for RPIT opaque types are unergonomic and require unobvious tricks.  The rules for capturing lifetime parameters are inconsistent with the rules for capturing lifetimes within type parameters.  The rules for RPIT are inconsistent with the rules for `async fn`, and this is exposed to users because of the common need to switch between these two equivalent forms.
+
+# Solution
+
+[solution]: #Solution
+
+This section is normative.
+
+In Rust 2024 and later editions, return position `impl Trait` (RPIT) opaque types will automatically capture all in-scope type *and* lifetime parameters.  In preparation for this, new RPIT-like `impl Trait` features introduced into the Rust 2021 edition will also automatically capture all in-scope type and lifetime parameters.
+
+## Apply `async fn` rule to RPIT in 2024 edition
+
+Under this RFC, in the Rust 2024 edition, RPIT opaque types will automatically capture all lifetime parameters in scope, just as `async fn` does in Rust 2021, and just as RPIT does in Rust 2021 when capturing type parameters.
+
+This updates and supersedes the behavior specified in [RFC 1522] and [RFC 1951].
+
+[RFC 1522]: https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md
+
+[RFC 1951]: https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md
+
+Consequently, the following examples will become legal in Rust 2024:
+
+### Capturing lifetimes from a free function signature
+
+```rust
+fn foo<'a, T>(x: &'a T) -> impl Sized { x }
+//                         ^^^^^^^^^^
+//                         ^ Captures `'a` and `T`.
+```
+
+### Capturing lifetimes from outer inherent impl
+
+```rust
+struct Foo<'a, T>(&'a T);
+impl<'a, T> Foo<'a, T> {
+    fn foo(self) -> impl Sized { self }
+    //              ^^^^^^^^^^
+    //              ^ Captures `'a` and `T`.
+}
+```
+
+### Capturing lifetimes from an inherent associated function signature
+
+```rust
+struct Foo<T>(T);
+impl<T> Foo<T> {
+    fn foo<'a>(x: &'a T) -> impl Sized { x }
+    //                      ^^^^^^^^^^
+    //                      ^ Captures `'a` and `T`.
+}
+```
+
+### Capturing lifetimes from an inherent method signature
+
+```rust
+struct Foo<T>(T);
+impl<T> Foo<T> {
+    fn foo<'a>(&self, x: &'a ()) -> impl Sized { (self, x) }
+    //                              ^^^^^^^^^^
+    // Captures `'_`, `'a`, and `T`.^
+}
+```
+
+## TAIT as the solution to overcapturing
```

Comment:
This probably shouldn't go in the normative section, maybe Rationale and Alternatives or Future Work would be better. I also agree that the text should mention https://github.com/rust-lang/rust/pull/116040.

Reply:
Agreed. I may be biased, but I think that rust-lang/rust#116040 is a more ergonomic solution to overcapturing than TAITs, too 😺

Reply:
Obviously this text was largely written before rust-lang/rust#116040 was conceived, and TAIT was the solution we had described normatively in the design meeting.  I had wanted to not touch any of the normative elements that had gained consensus in that meeting.  I'll give this some thought.  [This new section](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#rough-equivalence-between-opaques-with-and-without-a-specified-bound) in Appendix H presents the key idea that leads to how rust-lang/rust#116040 can be used to avoid the effects of overcapturing (but would need more explicit discussion to fully cover that particular point).

Reply:
In terms of mentioning rust-lang/rust#116040, note that [Appendix G: Future possibility: Inferred precise capturing](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#appendix-g-future-possibility-inferred-precise-capturing) does cover the problem that the PR addresses, and the precise description of what the effect of a solution would be was informed by the existence of that PR:

> Future work may relax this current limitation of the compiler.  The result of such an improvement would be that, when an outlives bound is specified for the opaque type, any type or lifetime parameters that the compiler could prove to not outlive that bound would act as if they were not captured by the opaque type.

Specifically, it informed the language "act as if they were not captured."  Under the PR, they are still captured, but the problem described in that section goes away.  There is a wrinkle about the fact that if something specifically checks for type equality, it does not actually "act as if", and I should probably cover that somehow (though it may be a bit of a tangent from the motivating examples in that section).

Reply:
After the discussion around https://github.com/rust-lang/rust/pull/116040 it's clear to me that it will become at least part of the solution, barring some unexpected problem. The purpose of TAIT in that design meeting was to show that there was an acceptable solution on the table for the overcapture problem, but even then, the potential need for a more ergonomic solution was discussed.

I think the RFC should keep these solutions on equal footing. https://github.com/rust-lang/rust/pull/116040 should provide a really ergonomic way to solve the problem in many, but perhaps not all, cases. TAIT provides a way to solve it in all cases, so we don't have to think too hard about whether we are making existing code impossible to write. To land this in 2024 we should still have TAIT (or if for some reason we can't get that, some _very compelling_ evidence that we are not reducing the expressiveness of the language in any real world scenarios).

Reply:
Regarding what was brought up in the meeting with respect to the potential need for a more ergonomic solution, note that the explicit captures syntax contemplated in the meeting is discussed in [Appendix F](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#appendix-f-future-possibility-precise-capturing-syntax).  We had also discussed the important case of a `+ 'static` bound working as it does today.  This is discussed in [Appendix E](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#appendix-e-adding-a-static-bound), and landing rust-lang/rust#116040 will eliminate any caveats about that (due to what is discussed in [Appendix G](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#appendix-g-future-possibility-inferred-precise-capturing)).

Regarding expressiveness limitations, as the document now points out:

> ...at the cost of an extra early-bound lifetime parameter in the signature of the callee, we can always express an RPIT without a specified outlives bound as an RPIT with a specified outlives bound in a way that does not change the requirements on the caller or the callee.  We do this by transforming a signature of the form `fn callee<P1, .., Pn>(..) -> impl Trait` to a signature of the form `fn callee<'o, P1: 'o, .., Pn: 'o>(..) -> impl Trait + 'o`.

Without TAIT, the expressiveness of the language is reduced in exactly two ways with the new capture rules and after landing (an idealized version of) rust-lang/rust#116040.

For functions that return an `impl Trait` opaque type that captures an in-scope lifetime parameter that was not previously captured, then 1) there is no way to restore type equality between two opaques resulting from invocations where different lifetimes were substituted, and 2) in the general case, we must add an extra early-bound lifetime parameter to the signature of our function to otherwise make the opaque type *act as if* the overcaptured lifetime was not captured, and this extra lifetime parameter visibly affects the API of our function.

For cases where we can add a `+ 'static` bound, or where one of the existing lifetime parameters suffices to express the desired bounds, then only the first limitation applies.

For rust-lang/rust#116040 as it exists rather than as idealized, there may be other (probably minor) expressiveness limitations due to the limitations mentioned in the PR:

> This approach has some limitations. Specifically, since liveness doesn't use the same type-test logic as outlives bounds do, we can't really try several options when we're faced with a choice.
>
> If we encounter two unique outlives regions in the param-env or bounds, we simply fall back to walking the opaque via its args. I expect this to be mostly mitigated by the special treatment of `'static`, and it can be fixed in a forwards-compatible by a more sophisticated analysis in the future.

(I'm planning to think more about everything above and about this discussion for a subsequent editing pass.)

Reply:
Thanks, that's a very helpful analysis.

To refine my earlier feedback, I do have a slight objection to the current structure which implies that we consider TAIT the one and only normative/intended solution, when it seems like the landscape has changed between now and when we had the design meeting. Now I would describe it as rust-lang/rust#116040 being the preferred solution for simple situations, with TAIT closing the remaining expressiveness gaps (and covering our bases in case we can't get rust-lang/rust#116040 for some reason).

For a minimal change, I would be fine with weakening the normative text to say something along the lines of "we need some solution to overcapturing and we agree that TAIT is sufficient for that purpose; additional solutions are discussed in the sections below". I think that captures what you're trying to say here.

Reply:
Agreed we're close here.  The relevant normative language in the document is:

> The stabilization of the 2024 lifetime capture rules in this RFC is contingent on the stabilization of TAIT or some other solution for precise capturing that will allow all code that is allowed under Rust 2021 to be expressed, in some cases with syntactic changes, in Rust 2024.

That is, it does say *"or some other solution"* in the same breath.  But I get what you mean about how we elsewhere lean on the word *"intended"*.  The way I think of it is that TAIT is the only one of the full solutions that has an approved RFC and that we seem to have consensus to stabilize in *some* form.  Other ideas missing those elements seem less currently *intended*.

It is an interesting question of to what degree landing rust-lang/rust#116040 might change our feelings about whether we would land the 2024 lifetime capture rules if a more full solution to overcapturing did not become available in time.  But that's a call we'd probably have to make then based on the situation.  In any case, I'd probably prefer to write another RFC to that effect rather than to extend the scope of this one.  (Simply noting more clearly in the document that this may have to be decided later is one approach I'm considering.)

Landing rust-lang/rust#116040 would be huge, and would absolutely mostly solve the overcapturing problem in practice when combined with the transformation described in [Appendix H](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#rough-equivalence-between-opaques-with-and-without-a-specified-bound).  Making this more of a theme elsewhere in the document is one potential revision I do have in mind.

(As mentioned, I'm planning to think more about this discussion and make a subsequent editing pass.)

Reply:
On the other hand, and to be very clear here, if we do believe that there is or could presently be a strong consensus that landing rust-lang/rust#116040 would entirely unblock landing the 2024 lifetime capture rules for RPIT in Rust 2024, then that would in fact simplify this RFC and make for a very straightforward revision.

Reply:
Agreed we're close here.  I'll reply back when an edit is ready.

Reply:
Note apropos of this discussion that a serious issue has been discovered with rust-lang/rust#116040.  If it turns out that we cannot land rust-lang/rust#116040, then the landscape will be the same as it was at the time of the 2023-07-26 design meeting (that is, rust-lang/rust#42940 was a known issue, and no solution to it was expected).

(I'll hold off on revisions that would have the document lean more heavily on rust-lang/rust#116040 until we better understand what is feasible.)

Reply:
The revisions discussed here have been made.  The document now more fully describes the landscape of solutions.  Discussion of TAIT has been removed from the normative section entirely, and the word "intended" has been purged from the document.

I'm going to mark this conversation as resolved as the revisions address the things discussed here, but of course, if there is more to add, please either unresolve the conversation or comment on the updated lines.